src/config/mongoose_config_parser_toml.erl

```diff
@@ -19,186 +30,127 @@ parse_file(FileName) ->
 process_general(<<"registration_timeout">>, V) ->
-    #local_config{key = registration_timeout, value = V};
+    [#local_config{key = registration_timeout, value = V}];
 process_general(<<"language">>, V) ->
-    #config{key = language, value = V};
+    [#config{key = language, value = V}];
 process_general(<<"all_metrics_are_global">>, V) ->
-    #local_config{key = all_metrics_are_global, value = V};
+    [#local_config{key = all_metrics_are_global, value = V}];
 process_general(<<"sm_backend">>, V) ->
-    #config{key = sm_backend, value = {b2a(V), []}};
+    [#config{key = sm_backend, value = {b2a(V), []}}];
 process_general(<<"max_fsm_queue">>, V) ->
-    #local_config{key = max_fsm_queue, value = V};
+    [#local_config{key = max_fsm_queue, value = V}];
 process_general(<<"http_server_name">>, V) ->
-    #local_config{key = cowboy_server_name, value = b2l(V)};
+    [#local_config{key = cowboy_server_name, value = b2l(V)}];
 process_general(<<"rdbms_server_type">>, V) ->
-    #local_config{key = rdbms_server_type, value = b2a(V)}.
-
-process_s2s_option(<<"use_starttls">>, V) ->
-    [#local_config{key = s2s_use_starttls, value = b2a(V)}];
-process_s2s_option(<<"certfile">>, V) ->
-    [#local_config{key = s2s_certfile, value = b2l(V)}];
-process_s2s_option(<<"default_policy">>, V) ->
-    [?HOST_F(#local_config{key = {s2s_default_policy, Host}, value = b2a(V)})];
-process_s2s_option(<<"outgoing_port">>, V) ->
-    [#local_config{key = outgoing_s2s_port, value = V}];
-process_s2s_option(<<"address">>, Addrs) ->
-    [#local_config{key = {s2s_addr, Host}, value = s2s_address(Addr)}
-     || Addr = #{<<"host">> := Host} <- Addrs].
-
-s2s_address(#{<<"ip_address">> := IP, <<"port">> := Port}) ->
-    {b2l(IP), Port};
-s2s_address(#{<<"ip_address">> := IP}) ->
-    b2l(IP).
+    [#local_config{key = rdbms_server_type, value = b2a(V)}].
 
+%% path: listen.*
+-spec process_listener_type(toml_key(), [toml_section()]) -> [option()].
 process_listener_type(Type, Content) ->
-    [process_listener(Type, C) || C <- Content].
+    parse_list(fun(L) -> process_listener(Type, L) end, Content).
 
+%% path: listen.*[]
+-spec process_listener(toml_key(), toml_section()) -> [option()].
 process_listener(Type, Content) ->
     Options = maps:without([<<"port">>, <<"ip_address">>], Content),
     PortIP = listener_portip(Content),
-    Opts = lists:flatmap(fun({K, V}) -> listener_opt(Type, K, V) end, maps:to_list(Options)),
+    Opts = parse_map(fun(K, V) -> listener_opt(Type, K, V) end, Options),
     {Port, IPT, _, _, Proto, OptsClean} =
         ejabberd_listener:parse_listener_portip(PortIP, Opts),
-    {{Port, IPT, Proto}, listener_module(Type), OptsClean}.
+    [{{Port, IPT, Proto}, listener_module(Type), OptsClean}].
 
+-spec listener_portip(toml_section()) -> option().
 listener_portip(#{<<"port">> := Port, <<"ip_address">> := Addr}) -> {Port, b2l(Addr)};
 listener_portip(#{<<"port">> := Port}) -> Port.
 
+-spec listener_module(toml_key()) -> option().
```

Comment:
Maybe we could include path here as well

Reply:
The idea was that when a path is the same as for the previous function, it is not specified again to avoid repetition.

In this case it would be still `listen.*[]` as we are still parsing the node representing the module - the argument is the key represented by the wildcard `*`. Actually this function could also be placed one level higher with `listen.*` as it only needs the type.

I think that I could fix this by explicitly stating the rule of no repetition in a comment. Would this s be ok?

Reply:
Yes, that is precisely what I had in mind